### PR TITLE
[aievec] Use intrinsics conversion to LLVM Dialect for matmul

### PR DESCRIPTION
**Pull request by jsetoain**

### CODE REVIEW

File: lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp [hunk 1]
```diff
@@ -921,76 +912,48 @@ class MatMulOpConversion
...
-          rewriter.getUnknownLoc(), intrinsicName,
-          LLVM::LLVMFunctionType::get(decodedMatMulOp.acc.getType(),
-                                      {decodedMatMulOp.lhs.getType(),
-                                       decodedMatMulOp.rhs.getType(),
-                                       decodedMatMulOp.acc.getType(), i32ty}));
+    else {
+      SmallVector<Type> intrFuncSig(
+          {VectorType::get({64}, rewriter.getI8Type()),
+           VectorType::get({16}, i32ty),
+           VectorType::get({16}, rewriter.getI64Type()), i32ty});
```
Comment (jamestcl-amd) [inline lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp:936]: Same here.

File: lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp [hunk 2]
```diff
@@ -921,76 +912,48 @@ class MatMulOpConversion
...
           rewriter
-              .create<LLVM::BitcastOp>(
+              .create<aievec::MacConfBF16IntrOp>(
                   loc, VectorType::get({8}, rewriter.getI64Type()),
-                  decodedMatMulOp.acc)
+                  forceCastOperandsToSignature(
+                      rewriter, loc, operands,
+                      {VectorType::get({32}, rewriter.getBF16Type()),
+                       VectorType::get({32}, rewriter.getBF16Type()),
+                       VectorType::get({8}, rewriter.getI64Type()), i32ty}))
```
Comment (jamestcl-amd) [inline lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp:930]: Is it possible to get the signature from the intrinsic op's definition?

File: test/Conversion/AIEVecToLLVM/matmul.mlir [hunk 3]
```diff
@@ -98,25 +84,24 @@ func.func @matmul(%A : vector<4x2xi32>, %B : vector<2x4xi16>,
...
+// CHECK:      %[[C0I32:.*]] = llvm.mlir.constant(0 : i32) : i32
+// CHECK:      %[[IFA:.*]] = "aievec.intr.set.I512.I256"(%[[IFA2512b]],
 // CHECK-SAME:                      %[[C0I32]]) : (vector<8xi32>, i32) ->
 // CHECK-SAME:                      vector<16xi32>
 // CHECK:      %[[BCA:.*]] = llvm.bitcast %[[IFA]] : vector<16xi32> to
 // CHECK-SAME:                      vector<64xi8>
 // CHECK:      %[[IFB2512b:.*]] = llvm.bitcast %[[FB]] : vector<8xi16> to
 // CHECK-SAME:                      vector<4xi32>
-// CHECK:      %[[IFB:.*]] = llvm.call @llvm.aie2.set.I512.I128(%[[IFB2512b]]) :
+// CHECK:      %[[IFB:.*]] = "aievec.intr.set.I512.I128"(%[[IFB2512b]]) :
```
[the remark this answers is not in the document]
Comment (jsetoain) [reply test/Conversion/AIEVecToLLVM/matmul.mlir:98]: Beware of internal links! The types are defined somewhere else, I'll point you to them.

File: lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp (hunk 2 above)
Comment (jsetoain) [reply lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp:930]: Not out of the box, no. The plan was to define an extra function that returns the signature, but it's not straightforward. I'll give it a go in the future.

**Review by jamestcl-amd — APPROVED** [review]

File: include/aie/Dialect/AIEVec/IR/AIEVecLLVMIntrOp.td [hunk 4]
```diff
@@ -46,6 +50,14 @@ def MacConfAcc32IntrOp :
                    VectorOfLengthAndType<[16], [I64]>:$acc,
                    I32:$conf)>;
 
+def MacConfAcc64IntrOp :
+    AIEVec2_IntrOp<"I512.I512.ACC1024.acc64.mac.conf",
+        [TypeIs<"res", VectorOfLengthAndType<[16], [I64]>>]>,
+    Arguments<(ins VectorOfLengthAndType<[64], [I8]>:$lhs,
+                   VectorOfLengthAndType<[16], [I32]>:$rhs,
+                   VectorOfLengthAndType<[16], [I64]>:$acc,
```
Comment (david-vc) [inline include/aie/Dialect/AIEVec/IR/AIEVecLLVMIntrOp.td:58]: Maybe a comment about why accumulators are represented as I64 would be in place?

**Review by david-vc — APPROVED** [review]
Hi @jsetoain :  Nice work! LGTM. David

File: include/aie/Dialect/AIEVec/IR/AIEVecLLVMIntrOp.td (hunk 4 above)
Comment (jsetoain) [reply include/aie/Dialect/AIEVec/IR/AIEVecLLVMIntrOp.td:58]: The same would apply to all the types. The explanation would make references to an internal project, so I'm being cautious here.

File: lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp [hunk 5]
```diff
@@ -13,19 +14,85 @@
...
+      .create<aievec::VectorSetI512I256IntrOp>(
+          loc, VectorType::get({16}, builder.getI32Type()),
+          bitcastValueToType(builder, loc, val,
+                             VectorType::get({8}, builder.getI32Type())),
+          builder.create<LLVM::ConstantOp>(loc, builder.getI32Type(),
+                                           (int32_t)0))
+      .getResult();
+}
+
+static Value forceCastValueToType(OpBuilder &builder, Location loc, Value val,
```
[the remark this answers is not in the document]
Comment (jsetoain) [reply lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp:55]: Indeed, all of these need to be documented. I'll take care of it once I can reproduce the test errors locally (or I give up and solve them blindly).

File: lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp [hunk 6]
```diff
@@ -13,19 +14,85 @@
...
+                                  Type type) {
+  auto valTy = val.getType();
+  if (valTy == type)
+    return val;
+  auto srcVecTy = dyn_cast<VectorType>(valTy);
+  if (srcVecTy) {
+    auto dstVecTy = dyn_cast<VectorType>(type);
+    assert(dstVecTy && "vector values cannot be forced into a non-vector type");
+    assert(srcVecTy.getRank() == 1 &&
```
[the remark this answers is not in the document]
Comment (jsetoain) [reply lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp:64]: 🤔 Can't hurt!

**Comment by jsetoain** [comment]
TIL: The order in which `clang` and `gcc` evaluate parameters in a function call is different.

**Pull request merged**